### PR TITLE
Added Compiler Tests. Added new utility functions.

### PR DESCRIPTION
**Pull request by hellofromtonya**
This PR adds tests for the Compiler API.  It also cleans up the tests, removing unnecessary code and complexity, especially in the integration tests.

It refactors the `beans_compile_{$type}_fragments()` functions, removing the redundant code and abstracting to it to a new private function called `_beans_compile_fragments()`.   This change makes the code DRY and centralizes how we handle the compiling into one function instead of 3.

It also adds 2 new utility functions:

- `beans_str_starts_with()`
- `beans_str_ends_with()`

These 2 new functions give developers and us some handy conditional checking for strings.  They also make the code more readable and understandable.

### CODE REVIEW

File: tests/phpunit/unit/api/compiler/beansCompileCssFragments.php [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+ * Class Test_BeansCompileCssFragments
+ *
+ * @package Beans\Framework\Tests\Unit\API\Compiler
+ * @group   api
+ * @group   api-compiler
+ */
+class Test_BeansCompileCssFragments extends Compiler_Test_Case {
+
+	/**
+	 * Test beans_compile_css_fragments() should return false when no fragments when no fragments are given.
```
Comment (christophherr) [inline tests/phpunit/unit/api/compiler/beansCompileCssFragments.php]: double `when no fragments`

File: tests/phpunit/unit/api/compiler/beansFlushCompiler.php [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+ * @package Beans\Framework\Tests\Unit\API\Compiler
+ * @group   api
+ * @group   api-compiler
+ */
+class Tests_BeansFlushCompiler extends Compiler_Options_Test_Case {
+
+	/**
+	 * Test beans_flush_compiler() should bail out when the directory does not exist.
+	 */
+	public function test_should_return_absolute_path_to_compiler_folder() {
```
Comment (christophherr) [inline tests/phpunit/unit/api/compiler/beansFlushCompiler.php]: DocBlock and function name mismatch.

File: tests/phpunit/unit/api/utilities/beansScandir.php [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+	protected function setUp() {
+		parent::setUp();
+
+		require_once BEANS_TESTS_LIB_DIR . 'api/utilities/functions.php';
+	}
+
+	/**
+	 * Test beans_scandir() should return false when directory is invalid.
+	 */
+	public function test_should_find_needle_when_array_given() {
```
Comment (christophherr) [inline tests/phpunit/unit/api/utilities/beansScandir.php]: DocBlock and function name mismatch

File: tests/phpunit/unit/api/utilities/beansScandir.php [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+		$this->assertFalse( beans_scandir( 'directory-does-not-exist' ) );
+	}
+
+	/**
+	 * Test beans_scandir() should return all files in given directory.
```
Comment (christophherr) [inline tests/phpunit/unit/api/utilities/beansScandir.php]: `in a given directory` or `in the given directory`

File: tests/phpunit/unit/api/utilities/beansStrEndsWith.php [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+		}
+
+		$this->assertFalse( beans_str_ends_with( false, 'ls' ) );
+		$this->assertTrue( beans_str_ends_with( true, '1' ) );
+		$this->assertFalse( beans_str_ends_with( false, '0' ) );
+		$this->assertTrue( beans_str_ends_with( false, '' ) );
+	}
+
+	/**
+	 * Test beans_str_ends_with() should correctly identify when given an array of needles.
```
Comment (christophherr) [inline tests/phpunit/unit/api/utilities/beansStrEndsWith.php]: when an array of needles is given.

File: tests/phpunit/unit/api/utilities/beansStrEndsWith.php [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+	 */
+	public function test_should_correctly_identify_when_given_an_array_of_needles() {
```
Comment (christophherr) [inline tests/phpunit/unit/api/utilities/beansStrEndsWith.php]: `when_array_of_needles_given`

File: tests/phpunit/unit/api/utilities/beansStrEndsWith.php [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+			[ 'rocks!', 'community', 'WordPress', 'the' ]
+		) );
+		$this->assertTrue( beans_str_ends_with(
+			'The WordPress Community Rocks!',
+			[ 'community rocks!', 'unity Rocks!' ]
+		) );
+	}
+
+	/**
+	 * Test beans_str_ends_with() should correctly identify when non-Latin given.
```
Comment (christophherr) [inline tests/phpunit/unit/api/utilities/beansStrEndsWith.php]: is given.

File: tests/phpunit/unit/api/utilities/beansStrEndsWith.php [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+	 * Prepares the test environment before each test.
+	 */
+	protected function setUp() {
+		parent::setUp();
+
+		require_once BEANS_TESTS_LIB_DIR . 'api/utilities/functions.php';
+	}
+
+	/**
+	 * Test beans_str_ends_with() should return false when case is wrong.
```
Comment (christophherr) [inline tests/phpunit/unit/api/utilities/beansStrEndsWith.php]: should this be
`when case does not match`?
(Similar for other tests here and in str_starts_width)

File: tests/phpunit/unit/api/utilities/beansStrStartsWith.php [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+	 * Prepares the test environment before each test.
+	 */
+	protected function setUp() {
+		parent::setUp();
+
+		require_once BEANS_TESTS_LIB_DIR . 'api/utilities/functions.php';
+	}
+
+	/**
+	 * Test beans_str_starts_with() should return false when case is wrong.
```
Comment (christophherr) [inline tests/phpunit/unit/api/utilities/beansStrStartsWith.php]: Please see comments in str_ends_with

File: tests/phpunit/unit/api/utilities/beansStrEndsWith.php [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+	/**
+	 * Test beans_str_ends_with() should return true when the case is correct.
+	 */
+	public function test_should_return_true_when_case_is_correct() {
+		$this->assertTrue( beans_str_ends_with( 'Foo', 'o' ) );
+		$this->assertTrue( beans_str_ends_with( 'Checking the Case?', 'Case?' ) );
+	}
+
+	/**
+	 * Test beans_str_ends_with() should correctly identify ending string pattern.
```
Comment (christophherr) [inline tests/phpunit/unit/api/utilities/beansStrEndsWith.php]: the ending string pattern.

File: tests/phpunit/unit/api/utilities/beansStrEndsWith.php [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+		$this->assertTrue( beans_str_ends_with( 'The WordPress Community Rocks!', 'Rocks!' ) );
+		$this->assertFalse( beans_str_ends_with( 'The WordPress Community Rocks!', 'rocks!' ) );
+		$this->assertTrue( beans_str_ends_with( 'This is a string test ', ' ' ) );
+		$this->assertFalse( beans_str_ends_with( 'This is a string test', ' ' ) );
+		$this->assertTrue( beans_str_ends_with( 'SomeClass::someMethod', 'd' ) );
+		$this->assertFalse( beans_str_ends_with( 'Fulcrum\Extender\Str\StrChecker::endsWith', '/' ) );
+	}
+
+	/**
+	 * Test beans_str_ends_with() should correctly identify when different data type is given.
```
Comment (christophherr) [inline tests/phpunit/unit/api/utilities/beansStrEndsWith.php]: a different data type

File: tests/phpunit/unit/api/utilities/beansStrEndsWith.php [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+	}
+
+	/**
+	 * Test beans_str_ends_with() should correctly identify when non-Latin given.
+	 */
+	public function test_should_correctly_identify_when_non_latin_given() {
+		$string = 'Τάχιστη αλώπηξ βαφής ψημένη γη, δρασκελίζει υπέρ νωθρού κυνός';
+
+		$this->assertFalse( beans_str_ends_with( $string, ' ' ) );
+		$this->assertFalse( beans_str_ends_with( $string, ' ' ) );
```
Comment (christophherr) [inline tests/phpunit/unit/api/utilities/beansStrEndsWith.php]: 114 and 115 are the same.

File: lib/api/compiler/class-beans-compiler.php [hunk 13]
```diff
@@ -732,8 +732,7 @@ private function get_new_hash( $hash, array $fragments_filemtime ) {
 	 * @return void
 	 */
 	private function remove_modified_files( $hash, $filemtime_hash ) {
-		$items = @scandir( $this->dir );  // phpcs:ignore Generic.PHP.NoSilencedErrors.Discouraged -- Valid use case.
-		unset( $items[0], $items[1] );
+		$items = beans_scandir( $this->dir );
```
Comment (christophherr) [inline lib/api/compiler/class-beans-compiler.php]: You used `beans_scandir` instead of `beans_scan_dir` because of the PHP `scandir` function?

Comment (hellofromtonya) [reply lib/api/compiler/class-beans-compiler.php]: Yes.  Keeping it similar and familiar.

Comment (christophherr) [reply lib/api/compiler/class-beans-compiler.php]: Makes sense 😄 👍

**Pull request merged**